js/gateio.js

```diff
@@ -2504,25 +2590,25 @@ module.exports = class gateio extends Exchange {
             });
         }
         return this.safeOrder ({
-            'id': id,
-            'clientOrderId': clientOrderId,
+            'id': this.safeNumber (order, 'id'),
+            'clientOrderId': this.safeNumber (order, 'user'),
             'timestamp': timestamp,
             'datetime': this.iso8601 (timestamp),
-            'lastTradeTimestamp': lastTradeTimestamp,
-            'status': status,
-            'symbol': symbol,
+            'lastTradeTimestamp': this.safeTimestamp2 (order, 'update_time', 'finish_time'),
+            'status': this.safeString (order, 'status'),
+            'symbol': this.safeSymbol (exchangeSymbol),
             'type': type,
             'timeInForce': timeInForce,
             'postOnly': undefined,
             'side': side,
-            'price': price,
-            'stopPrice': undefined,
-            'average': average,
-            'amount': amount,
+            'price': this.parseNumber (price),
+            'stopPrice': this.safeNumber (trigger, 'price'),
+            'average': this.safeNumber (order, 'fill_price'),
+            'amount': this.parseNumber (Precise.stringAbs (amount)),
             'cost': cost,
-            'filled': undefined,
-            'remaining': remaining,
-            'fee': undefined,
+            'filled': this.parseNumber (filled),
+            'remaining': this.parseNumber (Precise.stringAbs (remaining)),
+            'fee': this.safeNumber (order, 'fee'),
```

Comment:
We need to fix this line, the `fee` must be a dictionary, please, see how the `fee` substructure is filled in here:

- https://github.com/ccxt/ccxt/blob/master/js/huobi.js#L1770
- https://github.com/ccxt/ccxt/blob/master/js/huobi.js#L1783
- https://github.com/ccxt/ccxt/blob/master/js/ftx.js#L1015
- https://github.com/ccxt/ccxt/blob/master/js/mexc.js#L1036
- https://github.com/ccxt/ccxt/blob/master/js/okx.js#L1300

Reply:
Is that good?

Reply:
The fee field is there, I removed fees, it had both

Reply:
It should not be an array though. We should restore `fees`. We should fill in `fee` field with a single structure, if the exchange shows a single fee. The `fees` field should be filled with multiple fee-structures, if the exchange charges and shows fees in more than one currency on the same order/trade.

Reply:
Also, in the responses from your original message above, we should not fill the `fees` with both the maker fee and the taker fee. We should only fill the list of fees with the fees paid by "this user" on "this order", if the exchange charged it in multiple currencies.

Reply:
> on the fees, we generally implement the following approach... suppose you have a filled order
> the exchange could have charged the fees on that order in a single currency (say, USDT)
> or it could have charged the fees on that order in several currencies (part of the fee paid in BNB, another part in USDT)
> when the exchange only specifies a single value for the fee cost and a single currency – we must fill in the fee (single) field in the order
> in that case, the fees value will be calculated in safeOrder, so fees will become equal to [ fee ], and that will be returned to the user
>  when the exchange charges fees in multiple currencies, they can specify multiple fees in their responses, in that case we parse each fee into the fees array of fee-structures (multiple) and we leave the single fee undefined (edited)
> NEW
>  you can be charged in multiple currencies if , say, you have BNB fees enabled on Binance and you run out of BNB upon filling up an order, but still have USDT to cover the rest of the fee, in that case binance can charge the fees in different currencies, partially in each currency (edited)
>  so if the exchange specifies a single fee – we must set:
> 'fee': { // single fee
>     'cost': feeCost,
>     'currency': feeCurrencyCode,
>     // 'rate': '123.45', // only if available
> },
> 'fees': []

Reply:
> and if the exchange specified multiple fees we must set:
```
return this.safeOrder ({
    // ,..
    'fee': undefined,
    'fees': [ // multiple fees
        { 'cost': cost1, 'currency': code1, },
        { 'cost': cost2, 'currency': code2, },
        { 'cost': cost3, 'currency': code3, },
        // ...
    ],
    // ,..
}, market);
```
> both the fee and the fees property should only include the fees paid by this specific user
> the fees property should not include both the maker and the taker fees, cause, unless the user is both the taker and the maker on an order that crossed the midprice
> in other words, the fees property should not include other people's fees or counterparty fees 
> if you specify multiple fees of the same currency in the fees property, then safeOrder will take care of that and will sum them up and will fill a single fee field and will reduce the fees to their sum

Reply:
@kroitor I think I fixed it with 

```
        const numFeeCurrencies = fees.length;
        const multipleFeeCurrencies = numFeeCurrencies > 1;
        ...
            'fee': multipleFeeCurrencies ? undefined : fees,
            'fees': multipleFeeCurrencies ? fees : [],
```

Is `const multipleFeeCurrencies = numFeeCurrencies > 1;` good or do we want `const multipleFeeCurrencies = Precise.stringGt (numFeeCurrencies.toString(), '1');`?